Fix various issues with selections

labkey-alan opened the pull request:
#### Rationale
This PR fixes issues when users take actions with selections see issues:
- [Issue 53378](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=53378)
- [Issue 53797](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=53797)
- [Issue 53322](https://www.labkey.org/home/Developer/issues/issues-details.view?issueId=53322)

#### Related Pull Requests
- https://github.com/LabKey/labkey-ui-components/pull/1841
- https://github.com/LabKey/labkey-ui-premium/pull/801
- https://github.com/LabKey/limsModules/pull/1587
- https://github.com/LabKey/platform/pull/6988

#### Changes
- QueryModel.getSelectedIds: optimize when filterIds are not passed
- EntityMoveModal: Don't use selectionKey, simplify props
- Refactor PicklistEditModal, ChoosePicklistModal
  - Don't rely on selectionKey
  - Don't rely on QueryModel
- Add usePicklistSelections to get rid of duplicate selection loading logic
- Update AddToPicklistMenuItem, PicklistCreationMenuItem
- BulkUpdateForm
  - Remove props: header, onSubmitForEdit, sortString
  - Rename props: pluralNoun, singularNoun -> nounPlural, nounSingular
  - Add props: aliquots, sampleOperation, editStatusData
  - Render errors
- QueryAPIWrapper: remove getSelection
- EntityServerAPIWrapper.getCrossFolderSelectionResult: remove useSnapshotSelection argument

labkey-alan committed:
QueryModel.getSelectedIds: optimize when filterIds are not passed

labkey-alan committed:
EntityMoveModal: Don't use selectionKey, simplify props
- Partial Fix for Issue 53378

labkey-alan committed:
PicklistEditModal: Lint file

labkey-alan committed:
AddToPicklistMenuItem: Lint file

labkey-alan committed:
PicklistButton: Lint file

labkey-alan committed:
EntityMoveModal: Fix issue with undefined rowIds

labkey-alan committed:
PicklistCreationMenuItem.tsx - Lint file

labkey-alan committed:
Export LoadableState

labkey-alan committed:
Refactor PicklistEditModal, ChoosePicklistModal
- Don't rely on selectionKey
- Don't rely on QueryModel
- Add usePicklistSelections to get rid of duplicate selection loading logic
- Update AddToPicklistMenuItem, PicklistCreationMenuItem

labkey-alan committed:
picklist/actions.ts: Fix issue with getPicklistSamples

re-add error logging
refactor deletePicklists to use async request

labkey-alan committed:
BulkUpdateForm - Lint file

labkey-alan committed:
BulkUpdateForm - Remove unnecessary props, render warnings

All usages were using EntityBulkUpdateAlert, but that is in ui-premium. Code has been moved here so consumers don't need to manually wire up the warnings.

labkey-alan committed:
BulkUpdateForm: only select valid ids on load

labkey-alan committed:
BulkUpdateForm: fix issue with getSelectionCount

labkey-alan committed:
getOperationConfirmationData - actually use async/await

labkey-alan committed:
getOperationConfirmationDataForModel - Do not use selectionKey/useSnapshotSelection

labkey-alan committed:
PicklistEditModal: Fix title

labkey-alan committed:
getOperationNot(Allowed/Permitted)Message: add totalCount arg

- Necessary to handle deleted/missing selections

labkey-alan committed:
BulkUpdateForm: render error state when all selections are invalid

labkey-alan committed:
BulkUpdateForm: add additional test cases

labkey-alan committed:
BulkUpdateForm: add additional test cases

labkey-alan committed:
Remove getSelection

labkey-alan committed:
EntityServerAPIWrapper.getCrossFolderSelectionResult: remove useSnapshotSelection argument

labkey-alan committed:
AddToPicklistMenuItem / PicklistCreationMenuItem: Disable when there are no selected rows

labkey-alan committed:
getOperationNotPermittedMessage: make all arguments required

- Fix tests

labkey-alan committed:
PicklistEditModal: Fix issue with updating picklist, add error handling

Fix tests

labkey-alan committed:
getOperationNotAllowedMessageFromCounts: Better handle invalid counts

labkey-alan committed:
Add PicklistMenuItem: reduces duplicate code

Fix issue with PicklistCreationMenuItem being incorrectly disabled when acting as a button

labkey-alan committed:
BulkUpdateForm: Fix issue with editing not allowed values

labkey-nicka commented on packages/components/src/internal/actions.ts:
```diff
@@ -597,49 +597,6 @@ export function getSnapshotSelections(key: string, containerPath?: string): Prom
-            );
-        }
-
-        const params = getQueryParams(searchParams);
```
`getQueryParams` import can be removed.

labkey-nicka commented on packages/components/src/internal/components/entities/EntityMoveModal.tsx:
```diff
@@ -266,88 +212,70 @@ export const EntityMoveModal: FC<EntityMoveModalProps> = memo(props => {
     nounSingular: string,
-    nounPlural: string,
-    hideHelp?: boolean
+    nounPlural: string
```
nit: Return type of `message` can be `ReactNode` instead of `any`.

labkey-nicka commented on packages/components/src/internal/components/entities/actions.ts:
```diff
@@ -239,9 +233,7 @@ export async function getOperationConfirmationDataForModel(
     extraParams?: Record<string, any>
 ): Promise<OperationConfirmationData> {
-    const useSnapshotSelection = model.filterArray.length > 0;
-    if (useSnapshotSelection) await setSnapshotSelections(model.selectionKey, [...model.selections]);
```
Import of `setSnapshotSelections` can be removed.

labkey-nicka commented on packages/components/src/internal/components/entities/actions.ts:
```diff
@@ -239,9 +233,7 @@ export async function getOperationConfirmationDataForModel(
-    const useSnapshotSelection = model.filterArray.length > 0;
-    if (useSnapshotSelection) await setSnapshotSelections(model.selectionKey, [...model.selections]);
-    return getOperationConfirmationData(dataType, undefined, model.selectionKey, useSnapshotSelection, extraParams);
+    return getOperationConfirmationData(dataType, model.getSelectedIds(), undefined, undefined, extraParams);
```
Looks like the `getSelectedItemSamples()` in this file is no longer used and can be removed.

labkey-nicka commented on packages/components/src/internal/components/picklist/PicklistEditModal.test.tsx:
```diff
@@ -10,28 +10,34 @@ import { PRIVATE_PICKLIST_CATEGORY, PUBLIC_PICKLIST_CATEGORY } from './constants
 
 import { PicklistEditModal, PicklistEditModalProps } from './PicklistEditModal';
 import { Picklist } from './models';
+import { waitFor } from '@testing-library/dom';
```
nit: A couple of unused imports

labkey-nicka commented on packages/components/src/internal/components/picklist/PicklistEditModal.tsx:
```diff
@@ -217,38 +169,50 @@ const PicklistEditModalDisplay: FC<PicklistEditModalProps> = memo(props => {
+    const { error, loadingState, value: sampleIds } = useSampleSelections(selectedRowIds, sampleFieldKey, schemaQuery);
+    const loading = isLoading(loadingState);
+
+    if (loading || error !== undefined) {
```
Having a separate `Modal` instance from the one displayed later can result in flickering and odd component behavior under test. Consider consolidating loading behavior inside the `Modal` so that only one instance is created.

labkey-nicka commented on packages/components/src/internal/components/picklist/actions.ts:
```diff
@@ -2,9 +2,9 @@ import { Ajax, Domain, Filter, Query, Utils } from '@labkey/api';
-import { insertRows, QueryCommandResponse, selectRowsDeprecated } from '../../query/api';
-import { resolveKey, SchemaQuery } from '../../../public/SchemaQuery';
-import { getSelected, getSelectedDataDeprecated, setSnapshotSelections } from '../../actions';
+import { insertRows, QueryCommandResponse } from '../../query/api';
```
nit: Unused imports

labkey-nicka commented on packages/components/src/internal/components/picklist/actions.ts:
```diff
@@ -213,8 +195,8 @@ export function getSelectedPicklistSamples(
 ): Promise<number[]> {
-    return new Promise((resolve, reject) => {
-        getOrderedSelectedMappedKeys(
+    try {
```
No reason to try/catch here as `getOrderedSelectedMappedKeys()` and `setSnapshotSelections()` will already log the error to the console.

labkey-nicka commented on packages/components/src/internal/components/picklist/actions.ts:
```diff
@@ -301,93 +248,59 @@ export interface PicklistDeletionData {
-                reject(reason);
-            });
-    });
+export async function getPicklistDeleteData(model: QueryModel, user: User): Promise<PicklistDeletionData> {
```
Same about try/catch here (can be removed).

labkey-nicka commented on packages/components/src/internal/components/picklist/actions.ts:
```diff
@@ -223,74 +205,39 @@ export function getSelectedPicklistSamples(
+export async function addSamplesToPicklist(listName: string, sampleIds: number[]): Promise<QueryCommandResponse> {
+    const sampleIdsToAdd = await getSamplesNotInList(listName, sampleIds);
+    const rows = List(sampleIdsToAdd.map(id => ({ SampleId: id })));
+    const schemaQuery = new SchemaQuery('lists', listName);
```
Consider using a constant here `new SchemaQuery(SCHEMAS.PICKLIST_TABLES.SCHEMA, listName)`

labkey-nicka commented on packages/components/src/internal/components/picklist/actions.ts:
```diff
@@ -22,40 +22,29 @@ import { getOrderedSelectedMappedKeys } from '../entities/actions';
-        const { queryName, schemaName } = SCHEMAS.LIST_METADATA_TABLES.PICKLISTS;
-        selectRowsDeprecated({
+export async function getPicklistsForInsert(): Promise<Picklist[]> {
+    try {
```
Same about try/catch here.

labkey-nicka commented on packages/components/src/internal/components/picklist/usePIcklistSelections.ts:
```diff
@@ -0,0 +1,38 @@
+import { useCallback } from 'react';
```
Rename the file to `usePicklistSelections.ts`

labkey-alan replied:
lol

labkey-alan committed:
PR Feedback

labkey-alan committed:
PicklistEditModal: consolidate into single component

labkey-alan committed:
getCrossFolderSelectionResult - Remove dataRegionSelectionKey arg

AssayImportDropdownSection - Fix issue with crossfolder selections in subfolders

cnathe commented on packages/components/src/internal/components/forms/BulkUpdateForm.tsx:
```diff
@@ -187,11 +280,16 @@ export class BulkUpdateForm extends PureComponent<BulkUpdateFormProps, State> {
                         selected {noun}.
                     </p>
-                    {this.getSelectionCount() > 1 && onSubmitForEdit && (
-                        <p>To update individual {noun} in this selection group, select "Edit with Grid".</p>
```
I thought we had already taken out all of this "Edit with Grid" code? huh, must have missed some. Thanks for cleaning up.

cnathe commented on packages/components/src/internal/components/picklist/PicklistEditModal.tsx:
```diff
@@ -1,54 +1,49 @@
     sampleFieldKey?: string;
-    sampleIds?: number[];
+    schemaQuery?: SchemaQuery;
+    selectedRowIds?: number[] | string[];
```
I can't recall when we talked about this last, but don't we want these rowId related props to be number[] only and make the caller responsible for converting string to number as early as possible?

labkey-alan replied:
FreezerManager makes it really difficult to accomplish that, because it provides numbers. If I change FM to only pass strings I then also have to touch a lot of other downstream components to accept strings, which explodes the surface area of the changes, and I think would add unnecessary complexity to this PR, and potentially introduce subtle runtime issues.

labkey-alan committed:
getCrossFolderSelectionResult: improve return type

labkey-alan committed:
EntityMoveModal: warn about deleted selections

labkey-alan committed:
Prep for release
